src/test/regress/expected/with_modifying.out

```diff
@@ -120,8 +225,485 @@ WITH cte AS (
+
+SELECT * FROM summary_table ORDER BY id, counter;
+ id | counter 
+----+---------
+  1 |       1
+  1 |       2
+  2 |       1
+  2 |     154
+  3 |       1
+  3 |     185
+  4 |     160
+  5 |     170
+(8 rows)
+
+SELECT * FROM modify_table ORDER BY id, val;
+ id | val 
+----+-----
+(0 rows)
+
+SELECT * FROM anchor_table ORDER BY id;
+ id 
+----
+  1
+  2
+  3
+  4
+(4 rows)
+
+ROLLBACK;
+SELECT * FROM summary_table ORDER BY id, counter;
+ id | counter 
+----+---------
+  1 |       2
+  2 |     154
+  3 |     185
+  4 |     160
+  5 |     170
+(5 rows)
+
+SELECT * FROM modify_table ORDER BY id, val;
+ id | val 
+----+-----
+  1 |   2
+  2 |   4
+  3 |   6
+(3 rows)
+
+SELECT * FROM anchor_table ORDER BY id;
+ id 
+----
+  1
+  2
+(2 rows)
+
+-- Test delete with subqueries
+WITH deleted_rows AS (
+	DELETE FROM modify_table WHERE id IN (SELECT id FROM modify_table WHERE id = 1) RETURNING *
+)
+SELECT * FROM deleted_rows;
+ id | val 
+----+-----
+  1 |   2
+(1 row)
+
+WITH deleted_rows AS (
+	DELETE FROM modify_table WHERE id IN (SELECT id FROM modify_table WHERE val = 4) RETURNING *
+)
+SELECT * FROM deleted_rows;
+ERROR:  cannot perform distributed planning for the given modification
+DETAIL:  Joins are not supported in distributed modifications.
+WITH select_rows AS (
+	SELECT id FROM modify_table WHERE val = 4
+),
+deleted_rows AS (
+	DELETE FROM modify_table WHERE id IN (SELECT id FROM select_rows) RETURNING *
+)
+SELECT * FROM deleted_rows;
+ id | val 
+----+-----
+  2 |   4
+(1 row)
+
+WITH deleted_rows AS (
+	DELETE FROM modify_table WHERE val IN (SELECT val FROM modify_table WHERE id = 3) RETURNING *
+)
+SELECT * FROM deleted_rows;
+ERROR:  cannot perform distributed planning for the given modification
+DETAIL:  Joins are not supported in distributed modifications.
+WITH select_rows AS (
+	SELECT val FROM modify_table WHERE id = 3
+),
+deleted_rows AS (
+	DELETE FROM modify_table WHERE val IN (SELECT val FROM select_rows) RETURNING *
+)
+SELECT * FROM deleted_rows;
+ id | val 
+----+-----
+  3 |   6
+(1 row)
+
+WITH deleted_rows AS (
+	DELETE FROM modify_table WHERE ctid IN (SELECT ctid FROM modify_table WHERE id = 1) RETURNING *
+)
+SELECT * FROM deleted_rows;
+ERROR:  cannot perform distributed planning for the given modification
+DETAIL:  Joins are not supported in distributed modifications.
+WITH select_rows AS (
+	SELECT ctid FROM modify_table WHERE id = 1
+),
+deleted_rows AS (
+	DELETE FROM modify_table WHERE ctid IN (SELECT ctid FROM select_rows) RETURNING *
+)
+SELECT * FROM deleted_rows;
+ERROR:  cannot perform distributed planning for the given modification
+DETAIL:  Distributed modifications including system columns such as ctid on multiple shards are not supported.
+WITH added_data AS (
+	INSERT INTO modify_table VALUES (1,2), (1,6) RETURNING *
+),
+select_data AS (
+	SELECT * FROM added_data WHERE id = 1
+),
+raw_data AS (
+	DELETE FROM modify_table WHERE id = 1 AND ctid IN (SELECT ctid FROM select_data) RETURNING val
```

Comment:
hmz, we should not allow this but somehow the query is by-passing the check.

Looking at the query tree I see something like:
```
         {SUBPLAN 
         :subLinkType 2 
         :testexpr 
            {OPEXPR 
            :opno 387 
            :opfuncid 1292 
            :opresulttype 16 
            :opretset false 
            :opcollid 0 
            :inputcollid 0 
            :args (
               {VAR 
               :varno 1 
               :varattno -1 
               :vartype 27 
               :vartypmod -1 
               :varcollid 0 
               :varlevelsup 0 
               :varnoold 1 
               :varoattno -1 
               :location 203
```

Basically, the `read_intermediate_result` calls has been stripped from the rewritten query tree. In the multi-shard case we pass the original query to `ModifyQuerySupported`. 

I tried changing to `originalQuery` in `CreateModifyPlan`, but it seems that breaks some other tests, namely:

```sql 
 -- setting the partition column value to itself is allowed
 UPDATE limit_orders SET id = 246 WHERE id = 246;
+ERROR:  modifying the partition value of rows is not allowed
 UPDATE limit_orders SET id = 246 WHERE id = 246 AND symbol = 'GM';
+ERROR:  modifying the partition value of rows is not allowed

UPDATE example SET value = CASE WHEN true THEN now() ELSE now() + interval '1 hour' END WHERE key = 1;
+ERROR:  non-IMMUTABLE functions are not allowed in CASE or COALESCE statements
```

I guess we could maybe take the CTID check out of the `ModifyQuerySupported` function or pass original query as well.